application_sandbox/separate_depth_stencil_layouts/main.cpp

```diff
@@ -0,0 +1,716 @@
+    frame_data->framebuffer_ = containers::make_unique<vulkan::VkFramebuffer>(
+        data_->allocator(),
+        vulkan::VkFramebuffer(raw_framebuffer, nullptr, &app()->device()));
+
+    // Populate the render command buffer
+    frame_data->command_buffer_ =
+        containers::make_unique<vulkan::VkCommandBuffer>(
+            data_->allocator(), app()->GetCommandBuffer());
+
+    (*frame_data->command_buffer_)
+        ->vkBeginCommandBuffer((*frame_data->command_buffer_),
+                               &sample_application::kBeginCommandBuffer);
+    vulkan::VkCommandBuffer& cmdBuffer = (*frame_data->command_buffer_);
+
+    VkClearValue clears[3];
+    clears[0].color = {1.0f, 1.0f, 1.0f, 1.0f};
+    clears[1].depthStencil = {1.0f, 0};
+    clears[2].depthStencil = {1.0f, 0};
+
+    VkRenderPassBeginInfo pass_begin = {
+        VK_STRUCTURE_TYPE_RENDER_PASS_BEGIN_INFO,  // sType
+        nullptr,                                   // pNext
+        *torus_render_pass_,                       // renderPass
+        *frame_data->framebuffer_,                 // framebuffer
+        {{0, 0},
+         {app()->swapchain().width(),
+          app()->swapchain().height()}},  // renderArea
+        3,                                // clearValueCount
+        clears                            // clears
+    };
+
+    // Barrier before write to depth image
+    VkImageMemoryBarrier barrier{
+        VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER,
+        nullptr,
+        0,
+        0,
+        VK_IMAGE_LAYOUT_UNDEFINED,
+        VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL,
+        VK_QUEUE_FAMILY_IGNORED,
+        VK_QUEUE_FAMILY_IGNORED,
+        *frame_data->depth_stencil_image_,
+        {VK_IMAGE_ASPECT_DEPTH_BIT, 0, 1, 0, 1}};
+
+    cmdBuffer->vkCmdPipelineBarrier(cmdBuffer,
+                                    VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT,
+                                    VK_PIPELINE_STAGE_DRAW_INDIRECT_BIT, 0, 0,
+                                    nullptr, 0, nullptr, 1, &barrier);
+
+    // Draw the torus
+    cmdBuffer->vkCmdBeginRenderPass(cmdBuffer, &pass_begin,
+                                    VK_SUBPASS_CONTENTS_INLINE);
+    cmdBuffer->vkCmdBindDescriptorSets(
+        cmdBuffer, VK_PIPELINE_BIND_POINT_GRAPHICS,
+        ::VkPipelineLayout(*torus_pipeline_layout_), 0, 1,
+        &frame_data->torus_descriptor_set_->raw_set(), 0, nullptr);
+
+    cmdBuffer->vkCmdBindPipeline(cmdBuffer, VK_PIPELINE_BIND_POINT_GRAPHICS,
+                                 *torus_pipeline_);
+    cmdBuffer->vkCmdSetStencilReference(cmdBuffer, VK_STENCIL_FACE_FRONT_BIT,
+                                        0xAB);
+    cmdBuffer->vkCmdSetStencilWriteMask(cmdBuffer, VK_STENCIL_FACE_FRONT_BIT,
+                                        0x0F);
+    torus_.Draw(&cmdBuffer);
+
+    cmdBuffer->vkCmdEndRenderPass(cmdBuffer);
+
+    // Barrier before read from depth image
+    barrier.newLayout = VK_IMAGE_LAYOUT_DEPTH_READ_ONLY_OPTIMAL_KHR;
```

Comment:
Likewise we could do the implicit transition in the first subpass implicitly.

Reply:
Although this may be true, for the purpose of this sample I think using the barrier for at least this part is best, since that's a big part of what the extension provides.